server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +52,267 @@
+      final CoordinatorStats stats
   )
   {
-    CoordinatorStats stats = new CoordinatorStats();
+    final DruidCluster druidCluster = params.getDruidCluster();
 
     // Make sure we have enough loaded replicants in the correct tiers in the cluster before doing anything
-    for (Integer leftToLoad : loadStatus.values()) {
-      if (leftToLoad > 0) {
-        return stats;
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final String tier = entry.getKey();
+      // if there are replicants loading in cluster
+      if (druidCluster.hasTier(tier) && entry.getIntValue() > currentReplicants.getOrDefault(tier, 0)) {
+        return;
       }
     }
 
-    // Find all instances of this segment across tiers
-    Map<String, Integer> replicantsByTier = params.getSegmentReplicantLookup().getClusterTiers(segment.getIdentifier());
-
-    for (Map.Entry<String, Integer> entry : replicantsByTier.entrySet()) {
+    for (final Object2IntMap.Entry<String> entry : currentReplicants.object2IntEntrySet()) {
       final String tier = entry.getKey();
-      int loadedNumReplicantsForTier = entry.getValue();
-      int expectedNumReplicantsForTier = getNumReplicants(tier);
 
-      stats.addToTieredStat(DROPPED_COUNT, tier, 0);
+      final MinMaxPriorityQueue<ServerHolder> holders = druidCluster.getHistoricalsByTier(tier);
 
-      MinMaxPriorityQueue<ServerHolder> serverQueue = params.getDruidCluster().getHistoricalsByTier(tier);
-      if (serverQueue == null) {
-        log.makeAlert("No holders found for tier[%s]", entry.getKey()).emit();
-        continue;
+      final int numDropped;
+      if (holders == null) {
+        log.makeAlert("No holders found for tier[%s]", tier).emit();
+        numDropped = 0;
+      } else {
+        final int numToDrop = entry.getIntValue() - targetReplicants.getOrDefault(tier, 0);
+        numDropped = dropForTier(numToDrop, holders, segment);
       }
 
-      List<ServerHolder> droppedServers = Lists.newArrayList();
-      while (loadedNumReplicantsForTier > expectedNumReplicantsForTier) {
-        final ServerHolder holder = serverQueue.pollLast();
-        if (holder == null) {
-          log.warn("Wtf, holder was null?  I have no servers serving [%s]?", segment.getIdentifier());
-          break;
-        }
-
-        if (holder.isServingSegment(segment)) {
-          holder.getPeon().dropSegment(
-              segment,
-              null
-          );
-          --loadedNumReplicantsForTier;
-          stats.addToTieredStat(DROPPED_COUNT, tier, 1);
-        }
-        droppedServers.add(holder);
+      stats.addToTieredStat(DROPPED_COUNT, tier, numDropped);
+    }
+  }
+
+  private static int dropForTier(
+      final int numToDrop,
+      final MinMaxPriorityQueue<ServerHolder> holders,
+      final DataSegment segment
+  )
+  {
+    int numDropped = 0;
+
+    final List<ServerHolder> droppedHolders = new LinkedList<>();
```

Comment:
There is no point to use LinkedList here, please use ArrayList

Reply:
Sure, but won't a `LinkedList` will be more suitable here since the size is unknown and we don't need random access? Using a `LinkedList` will also prevent the problem of array resizing.

Reply:
No, if created as `ArrayList<>(1)` it is strictly always more memory efficient than LinkedList, and amortized cost of adding an element is smaller. `LinkedList` is almost never a good data structure as is (only intrusive linked lists are useful in some forms, sometimes). Exceptions, when LinkedList is useful itself, are so rare that I've never seen them in my practice, and there are no such in the Druid codebase.

Reply:
:+1: